File: C++/exercise/exercise_7.cpp
#include "exercise_7.h"
#include <iostream>

using namespace std;

vector<int> runLRUCache(int capacity, const vector<Query>& queries) {
    LRUCache cache(capacity);
    vector<int> results;
    
    cout << "Cache Size = " << capacity << endl;
    
    for (size_t i = 0; i < queries.size(); i++) {
        const Query& query = queries[i];
        
        if (query.operation == "SET") {
            cache.set(query.key, query.value);
            cout << "SET " << query.key << " " << query.value << " : ";
            cache.print();
            cout << endl;
        } else if (query.operation == "GET") {
            int result = cache.get(query.key);
            cout << "GET " << query.key << " : " << result;
            if (result != -1) {
                cout << " (";
                cache.print();
                cout << ")";
            }
            cout << endl;
            results.push_back(result);
        }
    }
    
    return results;
}

File: go/exercise/exercise_7.go
package exercises

import (
	"fmt"
	"laba2/structs/lru"
)

// Query представляет один запрос к LRU кэшу
type Query struct {
	operation string
	key       int
	value     int
}

// RunLRUCache выполняет серию запросов к LRU кэшу
func RunLRUCache(capacity int, queries []Query) []int {
	cache := lru.NewLRUCache(capacity)
	results := make([]int, 0)

	fmt.Printf("Инициализация LRU кэша с емкостью: %d\n", capacity)
	fmt.Println("========================================")

	for i, query := range queries {
		if query.operation == "SET" {
			fmt.Printf("Запрос %d: SET %d %d\n", i+1, query.key, query.value)
			cache.Set(query.key, query.value)
			cache.Print()
		} else if query.operation == "GET" {
			result := cache.Get(query.key)
			fmt.Printf("Запрос %d: GET %d -> %d\n", i+1, query.key, result)
			results = append(results, result)
			cache.Print()
		}
		fmt.Println("----------------------------------------")
	}

	return results
}

// CreateQuery создает новый запрос
func CreateQuery(operation string, key int, value int) Query {
	return Query{
		operation: operation,
		key:       key,
		value:     value,
	}
}
